### PR TITLE
Add: anonymise-attendance-log job

### PR DESCRIPTION
**Pull request by Leo310**
Resolving Issue: corona-school/project-user#505

Not tested and not sure how I should reset the pupilID.
And anonymsing the transaction log seems quite cumbersome, because as far as I could understand it, the courseAttendanceLog data is stringified and then stored in the transaction log.

### CODE REVIEW

File: jobs/periodic/anonymise-attendance-log/index.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import { getLogger } from "log4js";
+import { getManager } from "typeorm";
+import { CourseAttendanceLog } from "../../../common/entity/CourseAttendanceLog";
+
+const DELETE_AFTER = 7; // in days
+const logger = getLogger();
+
+export default async function execute() {
+    const entityManager = getManager();
+    await entityManager
```
Comment (Jonasdoubleyou) [inline jobs/periodic/anonymise-attendance-log/index.ts:10]: Prisma please :) 
And please also check that this does not also delete the pupil, the schema has a 'cascading' setting and I'm not sure what that does.

File: jobs/list.ts [hunk 2]
```diff
@@ -34,6 +35,7 @@ export const allJobs: CSCronJob[] = [
     // every 10 minutes during the day (to distribute load and send out notifications faster)
     { cronTime: '00 */10 * * * *', jobFunction: Notification.checkReminders },
     // each night - database cleanups
+    { cronTime: '00 00 05 * * *', jobFunction: anonymiseAttendanceLog },
```
Comment (Jonasdoubleyou) [inline jobs/list.ts:38]: If you alao add this to manualExecution.ts, you can run it via GraphQL with _executeJob, thats easier to test ...

File: jobs/periodic/anonymise-attendance-log/index.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+        .createQueryBuilder(CourseAttendanceLog, 'courseAttendanceLog')
+        .update()
+        .where('courseAttendanceLog.createdAt < :date', { date: Date.now() - DELETE_AFTER })
+        .set({ ip: "", pupil: { id: 0 } })
```
Comment (Jonasdoubleyou) [inline jobs/periodic/anonymise-attendance-log/index.ts:14]: So here the problem is that the pupilId is non nullable?

Comment (Leo310) [reply jobs/periodic/anonymise-attendance-log/index.ts:14]: So what pupilId should I then reference?
image

Should it point to an existing empty Mustermann pupil?

Comment (lsc0498) [reply jobs/periodic/anonymise-attendance-log/index.ts]: Yes, that's fine. There is no need to anonymize the (internal) pupilId. As long as no IP address, names or email addresses are left if a user deactivates their account it should be okay 👍

File: jobs/periodic/anonymise-attendance-log/index.ts [hunk 4]
```diff
@@ -0,0 +1,10 @@
+import { prisma } from "../../../common/prisma";
+import { getLogger } from "log4js";
+
+const DELETE_AFTER = 7; // in days
+const logger = getLogger();
+
+export default async function execute() {
+    await prisma.course_attendance_log.updateMany({ where: { createdAt: { lt: new Date(Date.now() - DELETE_AFTER) } }, data: { ip: "" } });
```
Comment (Jonasdoubleyou) [inline jobs/periodic/anonymise-attendance-log/index.ts:8]: Date.now() is in milliseconds, - 7 is 7 milliseconds before

**Review by Jonasdoubleyou — CHANGES_REQUESTED** [review]

**Comment by Leo310** [comment]
"And anonymsing the transaction log seems quite cumbersome, because as far as I could understand it, the courseAttendanceLog data is stringified and then stored in the transaction log."

Any thoughts on this point?

**Review by Jonasdoubleyou — DISMISSED** [review]

**Comment by Leo310** [comment]
Bisherige logs müssten trotzdem nochmal mit scripts gecleaned und anonymisiert werden.

**Review by Jonasdoubleyou — APPROVED** [review]

**Pull request merged**